### PR TITLE
[13608] Dynamic memory topic payload pool properly handle Zero sized payloads

### LINKED ISSUE #1929: DYNAMIC_REUSABLE_MEMORY_MODE breaks new subscriber reusing an existing participant [13608]
If I have a publisher that sets `ParticipantAttributes::rtps::builtin::readerHistoryMemoryPolicy` to `DYNAMIC_REUSABLE_MEMORY_MODE`, and I have a subscriber that is destroyed and recreated with a reused participant, the subscriber never receives any new data from that publisher.

## Expected Behavior
The subscriber should start to receive data again.

## Current Behavior
The subscriber stops receiving data.

## Steps to Reproduce
<details>
<summary>rtpsBool.idl</summary>

```
struct rtpsBool_t
{
	boolean value;
};
```
</details>

<details>
<summary>main.cpp</summary>

```cpp
#include "rtpsBool.h"
#include "rtpsBoolPubSubTypes.h"

#include <fastrtps/Domain.h>
#include <fastrtps/attributes/ParticipantAttributes.h>
#include <fastrtps/publisher/Publisher.h>
#include <fastrtps/subscriber/SampleInfo.h>
#include <fastrtps/subscriber/Subscriber.h>
#include <fastrtps/subscriber/SubscriberListener.h>

using namespace eprosima::fastrtps;
using namespace eprosima::fastrtps::rtps;

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <getopt.h>
#include <memory>
#include <thread>
using namespace std::chrono_literals;

namespace
{
constexpr auto TOPIC_NAME = "test";

using participant_t = std::unique_ptr<Participant, bool (*) (Participant *)>;
using publisher_t   = std::unique_ptr<Publisher, bool (*) (Publisher *)>;
using subscriber_t  = std::unique_ptr<Subscriber, bool (*) (Subscriber *)>;

class listener_t : public SubscriberListener
{
public:
	void onNewDataMessage (Subscriber *const sub_) override
	{
		rtpsBool_t data;
		SampleInfo_t info;

		sub_->takeNextData (&data, &info);
		if (info.sampleKind != eprosima::fastrtps::rtps::ALIVE)
			return;

		std::printf ("%s: %s\n", __func__, data.value () ? "true" : "false");
	}
};

rtpsBool_tPubSubType pubSubType;

participant_t makeParticipant (bool const dynamic_)
{
	auto attr = std::make_unique<ParticipantAttributes> ();

	if (dynamic_)
		attr->rtps.builtin.readerHistoryMemoryPolicy = rtps::DYNAMIC_REUSABLE_MEMORY_MODE;

	return participant_t (Domain::createParticipant (*attr), &Domain::removeParticipant);
}

publisher_t makePublisher (participant_t &participant_)
{
	auto attr = std::make_unique<PublisherAttributes> ();

	attr->topic.topicDataType = pubSubType.getName ();
	attr->topic.topicName     = TOPIC_NAME;

	return publisher_t (Domain::createPublisher (participant_.get (), *attr), &Domain::removePublisher);
}

subscriber_t makeSubscriber (participant_t &participant_, listener_t *const listener_)
{
	auto attr = std::make_unique<SubscriberAttributes> ();

	attr->topic.topicDataType = pubSubType.getName ();
	attr->topic.topicName     = TOPIC_NAME;

	return subscriber_t (Domain::createSubscriber (participant_.get (), *attr, listener_), &Domain::removeSubscriber);
}
}

int main (int argc_, char *argv_[])
{
	bool dynamic   = false;
	bool publisher = false;

	int c;
	while ((c = getopt (argc_, argv_, "dp")) != -1)
	{
		if (c == 'd')
			dynamic = true;
		else if (c == 'p')
			publisher = true;
	}

	listener_t listener;

	auto participant = makeParticipant (dynamic);
	if (!Domain::registerType (participant.get (), &pubSubType))
	{
		std::printf ("Failed to register type\n");
		return EXIT_FAILURE;
	}

	if (publisher)
	{
		rtpsBool_t data;
		data.value (false);

		std::printf ("Creating publisher\n");
		auto pub = makePublisher (participant);

		while (true)
		{
			data.value (!data.value ());
			std::printf ("Publishing %s\n", data.value () ? "true" : "false");
			pub->write (&data);
			std::this_thread::sleep_for (500ms);
		}
	}
	else
	{
		subscriber_t sub = std::unique_ptr<Subscriber, bool (*) (Subscriber *)> (nullptr, nullptr);

		// alternate creating and destroying subscriber, reusing the existing participant
		while (true)
		{
			if (!sub)
			{
				std::printf ("Creating subscriber\n");
				sub = makeSubscriber (participant, &listener);
				std::this_thread::sleep_for (3s);
			}
			else
			{
				std::printf ("Destroying subscriber\n");
				sub.reset ();
				std::this_thread::sleep_for (1s);
			}
		}
	}
}
```
</details>

1. Run subscriber: `./program`
2. In another terminal, run publisher:
  a. Succeeds: `./program -p`
  b. Fails: `./program -pd`

## System information
- Fast-RTPS version: 2.2.0, 2.3.0 (problem didn't occur with 1.9.4)
- OS: CentoOS 7, Fedora 34, Ubuntu 20.04
- Network interfaces: lo
- ROS2: N/A, just using fastrtps directly.

## Additional context
The problem does not occur if you create the subscriber with a brand new participant. We share and reuse participants gratuitously to avoid using too many sockets and having too much metatraffic.

If using discovery server, the problem occurs once. So the first created subscriber receives data, the second created subscriber doesn't received data, and then all subsequently created subscribers do receive data.

We were previously running with 1.9.4. Since trying to upgrade to 2.2.0 we have run into this problem and this is the minimum test case I could make to reproduce it. We will use the default readerHistoryMemoryPolicy in the meantime.

### PR DESCRIPTION
**Pull request by Mario-DL**
As a regression of #1929, this PR introduces a couple of unit tests, the regression BB test and proposes a fix for accepting zero sized payload allocations in the topic pool. 

In particular, this is interesting to properly add to the history the disposals of any endpoints in a remote Participant configured with DYNAMIC_*_MEMORY modes, as the payload size in that case is 0. Otherwise, the disposal CacheChanges are always discarded by remote participants and are not added to the history.

*Note: this PR should not be merged until 2.11 freeze is finished*

## Description

@Mergifyio backport 2.11.x 2.10.x 2.9.x 2.6.x

Fixes #1929

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [X] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by Mario-DL** [comment]
@richiprosima please test this

File: test/blackbox/api/dds-pim/PubSubWriter.hpp [hunk 1]
```diff
@@ -645,6 +645,35 @@ class PubSubWriter
...
+            unsigned int matched = 0)
+    {
+        bool ret_value = true;
+        std::unique_lock<std::mutex> lock(mutexDiscovery_);
+
+        std::cout << "Writer is waiting removal..." << std::endl;
+
+        if (!cv_.wait_for(lock, timeout, [&]()
+                {
+                    return matched_ == matched;
```
Comment (EduPonz) [inline test/blackbox/api/dds-pim/PubSubWriter.hpp:659]:
```suggestion
                    return matched_ <= matched;
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Review by EduPonz — APPROVED** [review]

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Comment by EduPonz** [comment]
https://github.com/Mergifyio backport 2.11.x 2.10.x 2.9.x 2.6.x

**Pull request merged**